### PR TITLE
Add mapPartitionsWithIndex

### PR DESCRIPTION
**Pull request by holdenk**
Add mapPartitionsWithIndex

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/api/java/JavaRDDLike.scala [hunk 1]
```diff
@@ -68,6 +68,15 @@ trait JavaRDDLike[T, This <: JavaRDDLike[T, This]] extends Serializable {
     new JavaRDD(rdd.map(f)(f.returnType()))(f.returnType())
 
   /**
+   * Return a new RDD by applying a function to each partition of this RDD, while tracking the index
+   * of the original partition.
+   */
+  def mapPartitionsWithIndex[R: ClassManifest](f: JFunction2[Int, java.util.Iterator[T], java.util.Iterator[R]],
+			     preservesPartitioning: Boolean = false): JavaRDD[R] =
```
Comment (rxin) [inline core/src/main/scala/org/apache/spark/api/java/JavaRDDLike.scala]: Thanks. Can we change this to indent 4 characters? 

```
def mapPartitionsWithIndex[R: ClassManifest](
    f: JFunction2[Int, java.util.Iterator[T], java.util.Iterator[R]],
    preservesPartitioning: Boolean = false): JavaRDD[R] =
```

**Comment by holdenk** [comment]
Done :)

**Comment by rxin** [comment]
Sorry there was a misunderstanding. We use 2 space indenting in general, but for function declarations in which parameters don't fit in a single line, we use 4 spaes.

See the example I gave.

**Comment by holdenk** [comment]
Fixed again (but don't see the example?)

**Comment by rxin** [comment]
I've merged this into asf spark branch-0.8.

File: core/src/main/scala/org/apache/spark/api/java/JavaRDDLike.scala [hunk 2]
```diff
@@ -68,6 +68,16 @@ trait JavaRDDLike[T, This <: JavaRDDLike[T, This]] extends Serializable {
     new JavaRDD(rdd.map(f)(f.returnType()))(f.returnType())
 
   /**
+   * Return a new RDD by applying a function to each partition of this RDD, while tracking the index
+   * of the original partition.
+   */
+  def mapPartitionsWithIndex[R: ClassManifest](
```
Comment (JoshRosen) [inline core/src/main/scala/org/apache/spark/api/java/JavaRDDLike.scala:74]: I'm think that declaring `[R:ClassManifest]` introduces an implicit ClassManifest argument to this function, which will be problematic when calling it from Java.  Can you fix this and add a unit test to JavaRDDSuite to verify that this method is callable from Java without having to pass the manifests?

**Pull request merged**